src/dataflow/src/sink/kafka.rs

```diff
@@ -574,3 +581,161 @@ where
+
+    let mut stash: HashMap<Capability<Timestamp>, Vec<_>> = HashMap::new();
+    let mut vector = Vec::new();
+    let mut encode_logic = move |input: &mut FrontieredInputHandle<
+        Timestamp,
+        ((Option<Row>, Option<Row>), Timestamp, Diff),
+        _,
+    >,
+                                 output: &mut OutputHandle<
+        _,
+        ((Option<Vec<u8>>, Option<Vec<u8>>), Timestamp, Diff),
+        _,
+    >| {
+        let mut fuel_remaining = fuel;
+        // stash away all the input we get, we want to be a nice citizen
+        input.for_each(|cap, data| {
+            data.swap(&mut vector);
+            let stashed = stash.entry(cap.retain()).or_default();
+            for record in vector.drain(..) {
+                stashed.push(record);
+            }
+        });
+
+        // work off some of our data and then yield, can't be hogging
+        // the worker for minutes at a time
+
+        while fuel_remaining > 0 && !stash.is_empty() {
+            let lowest_ts = stash
+                .keys()
+                .min_by(|x, y| x.time().cmp(y.time()))
+                .expect("known to exist")
+                .clone();
+            let records = stash.get_mut(&lowest_ts).expect("known to exist");
+
+            let mut session = output.session(&lowest_ts);
+            let num_records_to_drain = cmp::min(records.len(), fuel_remaining);
+            // let num_records_to_drain = records.len();
+            records
+                .drain(..num_records_to_drain)
+                .for_each(|((key, value), time, diff)| {
+                    let key =
+                        key.map(|key| encoder.encode_key_unchecked(key_schema_id.unwrap(), key));
+                    let value =
+                        value.map(|value| encoder.encode_value_unchecked(value_schema_id, value));
+                    session.give(((key, value), time, diff));
+                });
+
+            fuel_remaining -= num_records_to_drain;
+
+            if records.is_empty() {
+                // drop our capability for this time
+                stash.remove(&lowest_ts);
+            }
+        }
+
+        if !stash.is_empty() {
+            activator.activate();
+            return true;
+        }
+        // signal that we're still incomplete
+        !stash.is_empty()
+    };
+
+    builder.build_reschedule(|_capabilities| {
+        move |frontiers| {
+            let mut output_handle = output.activate();
+            let mut input_handle = FrontieredInputHandle::new(&mut input, &frontiers[0]);
+            encode_logic(&mut input_handle, &mut output_handle)
+        }
+    });
+
+    output_stream
+}
+
+/// Pulls all data to one worker by hashing on the sink id. Sorts and emits data
+/// that is before the frontier.
+fn sort_on_one_worker<G>(
+    input_stream: &Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>,
+    sink_name: String,
+    sink_id: GlobalId,
+) -> Stream<G, ((Option<Vec<u8>>, Option<Vec<u8>>), u64, isize)>
```

Comment:
This seems like a weird timely dataflow operator, from personal experience. It could just be evolution from our code base, but: generally, each operator has the ability to drive its own data distribution rules, and if it needs its input to be e.g. on one worker, it should do that (to make sure that the contract is maintained). 

If it turns out that the problem is that too many operators need this behavior, we can try and wrap it up. For example, timely's `Notificator` struct wraps input streams and provide their contents in time order.

Reply:
Like above, this would be a great place to put more doccomments about what the method assumes about its input, and what it guarantees about its output. For example, what are the data sorted by, and how should a user use the output of this method (I would personally put big red warning flags that the ordering is potentially destroyed if a user invokes an operator that exchanges the data, which is one of the reasons I am a bit weirded out about it).

Reply:
Agreed! What got me started on this is that the Kafka sink operator was assuming that `render/sinks.rs` would a) sort the data by timestamp, and b) exchange it all to one operator. This assumption is incorrect so I moved that responsibility "closer" to the Kafka operator.

I think I didn't move it close enough and should include the "exchanging all to one operator and sort it " part into the Kafka sink operator (via its Pact and then by sorting internally).

I will do that now, but I also need to think about future Kafka Sinks that don't need this sorting/exchanging to one operator. Thinking about a CDCv2 sink here. It might turn out that the existing Kafka Sink and that one can't share much because their requirements are too different.